File: html/search_page.rb
require 'interface_database.rb',
        'interface_list_item.rb',
        'class_page.rb',
        'page.rb' do

class SearchPage < Page

  attr_accessor :search_text
  attr_reader :interfaces

  def initialize(url: nil)
    super('search')
    self.search_text = url.split('/')[0] if url
    @interface_list_items = []
  end

  def location_bar_url
    return "/#{@page_name}" if !search_text || search_text.size == 0
    return "/#{@page_name}/#{search_text}"
  end

  def search_text
    if !@input
      @initial_search_text 
    else
      @input.value
    end
  end

  def search_text=(new_text)
    if !@input
      @initial_search_text = new_text
    else
      @input.value = new_text
      on_search_change(nil)
    end
  end

  def did_appear
    @input.focus
  end

  private

  def did_load
    @input = @element.query_selector('#search-input')
    @result_list = @element.query_selector('.result-list')
    @header = @element.query_selector('header')

    @input.add_event_listener('keyup', method(:on_search_change))

    if @initial_search_text
      @input.value = @initial_search_text
      on_search_change(nil)
      remove_instance_variable(:@initial_search_text)
    end
  end

  def on_search_change(event)
    search_string = @input.value
    WebDocs.page_stack.update_location_bar_url
    $window.console.log("Search changed, input=#{search_string}")

    InterfaceDatabase.instance.find_interfaces(search_string) do |interfaces|
      next if search_string != @input.value
      self.interfaces = interfaces
    end
  end

  def interfaces=(new_interfaces)
    @interfaces = new_interfaces

    if !@interfaces
      @interface_list_items.each { |item| item.style.display = 'none' }
      return
    end

    @interfaces.each_with_index do |interface, index|
      item = @interface_list_items[index]
      if !item
        item = InterfaceListItem.new
        item.show_owner_class = true
        @interface_list_items.push(item)
        @result_list.append_child(item)
      end

      item.interface = interface
      item.class_list.remove('display-none')
    end

    items_to_none = @interface_list_items[@interfaces.size..@interface_list_items.size]
    items_to_none.each do |item|
      break if item.class_list.contains('display-none')
      item.class_list.add('display-none')
    end
  end

  def on_click_interface(event)
    interface = event.detail.interface
    class_page = ClassPage.new
    if interface[:interface_type] == :class
      class_page.interface = interface
    else
      class_page.interface = interface[:owner]
      class_page.current_member = interface
    end
    
    WebDocs.page_stack.push(page: class_page, animated: true)
  end

end # SearchPage

end # require

File: site/scripts/search-page.js
var SearchPage = function(url) {
  Page.call(this, 'search');
  this._interfaceListItems = [];

  if (url) {
    var escapedSearchText = url.split('/')[0];
    this.searchText = decodeURIComponent(escapedSearchText);
  }
};

SearchPage.prototype = new Page();

Object.defineProperty(SearchPage.prototype, 'searchText', {
  get: function() {
    if (!this._input) {
      return this._initialSearchText;
    } else {
      return this._input.value;
    }
  },
  set: function(newValue) {
    if (!this._input) {
      this._initialSearchText = newValue;
    } else {
      this._input.value = newValue;
      this._onSearchChange(null);
    }
  }
});

Object.defineProperty(SearchPage.prototype, 'interfaces', {
  get: function() {
    return this._interfaces;
  },
  set: function(newInterfaces) {
    this._interfaces = newInterfaces;

    if (!this._interfaces) {
      this._interfaceListItems.forEach(function(item) {
        item.element.style.display = 'none';
      });
      return;
    }

    this._interfaces.forEach(function(anInterface, index) {
      var item = this._interfaceListItems[index];
      if (!item) {
        item = new InterfaceListItem();
        item.showOwnerClass = true;
        item.element.addEventListener(InterfaceListItem.CLICKED_INTERFACE, this._onClickInterface);
        this._interfaceListItems.push(item);
        this._resultList.appendChild(item.element);
      }

      item.interfaceData = anInterface;
      item.element.classList.remove('display-none');
    }, this);

    var itemsToNone = this._interfaceListItems.slice(this._interfaces.length);
    for (var i = 0; i < itemsToNone.length; i++) {
      var item = itemsToNone[i];
      if (item.element.classList.contains('display-none')) {
        break;
      }
      item.element.classList.add('display-none');
    }
  }
});

SearchPage.prototype.locationBarUrl = function() {
  if (!this.searchText || this.searchText.length === 0) {
    return '/' + this.pageName;
  }

  return '/' + this.pageName + '/' + this.searchText;
};

SearchPage.prototype.pageTitle = function() {
  if (!this.searchText || this.searchText.length === 0) {
    return 'Webster: Reference for the web';
  }

  return '"' + this.searchText + '": Webster';
};

SearchPage.prototype._didAppear = function() {
  this._input.focus();
};

SearchPage.prototype._didLoad = function() {
  this._input = this.element.querySelector('#search-input');
  this._resultList = this.element.querySelector('.result-list');
  this._header = this.element.querySelector('header');

  this._input.addEventListener('keyup', this._onSearchChange.bind(this));

  if (this._initialSearchText) {
    this.searchText = this._initialSearchText;
    delete this._initialSearchText;
  }

  this.element.querySelector('.info-button').onclick = this._onClickInfo;
};

SearchPage.prototype._onClickInfo = function(event) {
  // Allow them to open the link in a new window or tab.
  if (event.ctrlKey || event.metaKey) {
    return;
  }

  event.preventDefault();
  var infoPage = new InfoPage();
  WebDocs.pageStack.push(infoPage, true);
};

SearchPage.prototype._onSearchChange = function(event) {
  var searchString = this._input.value;
  WebDocs.pageStack.updateLocationBarUrl();
  WebDocs.pageStack.updatePageTitle();

  InterfaceDatabase.findInterfaces(searchString, function(foundInterfaces) {
    if (searchString != this._input.value) {
      return;
    }

    this.interfaces = foundInterfaces;
  }.bind(this));
};

SearchPage.prototype._onClickInterface = function(event) {
  var interfaceData = event.detail.interfaceData;
  var classPage = new ClassPage();
  if (interfaceData.interfaceType == 'class') {
    classPage.interfaceData = interfaceData;
  } else {
    classPage.interfaceData = interfaceData.owner;
    classPage.currentMember = interfaceData;
  }

  WebDocs.pageStack.push(classPage, true);
};